Convert JavascriptString::EntryLocaleCompare to use LogicalStringCompare

msdimitrov opened the pull request:
Resolves #4841
Resolves #4843 

I didn't change the error message, since I'm not sure if a new error message should be added to `rterros.h` or throw the error from `GetLastError()` .

jackhorton commented on lib/Runtime/PlatformAgnostic/Platform/Windows/UnicodeText.cpp:
```diff
@@ -416,9 +416,19 @@ namespace PlatformAgnostic
+            // CompareStringEx called with these flags is equivalent to calling StrCmpLogicalW
             // and we have the added advantage of not having to link with shlwapi.lib just for one function
-            int i = CompareStringW(LOCALE_USER_DEFAULT, NORM_IGNORECASE | SORT_DIGITSASNUMBERS, string1, -1, string2, -1);
+            int i = CompareStringEx(localeName, NORM_IGNORECASE | SORT_DIGITSASNUMBERS, string1, -1, string2, -1, NULL, NULL, 0);
```
I believe you can pass either `""` or `nullptr` for the user default locale (not both -- one meant user default and the other meant invariant). That should make it so you don't have to call `GetUserDefaultLocaleName`.

jackhorton commented on lib/Runtime/Library/JavascriptString.cpp:
```diff
@@ -1358,30 +1360,21 @@ namespace Js
         const char16* pThatStr = pThat->GetString();
-        int thatStrCount = pThat->GetLength();
+        
+        int result = UnicodeText::LogicalStringCompare(pThisStr, pThatStr);
```
Strings can have embedded nulls in JS, so we will need to change this API to pass in the actual length of each string, rather than assuming null termination (especially since GetString is not required to actually give back a null-terminated buffer -- that's what GetSz is for). Since we obviously don't have a test for this case, it would be great to add one!

MSLaguana commented on test/Strings/compare.js:
```diff
@@ -9,10 +9,12 @@ var str1 = "abcd1234"
+var str1null = "\0ab\0cd1234\0";
 
 assert.isTrue(str1.localeCompare(str1a) == 0);
+assert.isTrue(str1.localeCompare(str1null) == 0);
```
Seems like this is failing on non-ICU versions of chakracore, it seems that winglob doesn't ignore nulls in comparisons?

msdimitrov replied:
Correct. I'm currently changing the the function to take the length of the strings.

jackhorton commented on lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.ICU.cpp:
```diff
@@ -266,9 +266,9 @@ namespace PlatformAgnostic
-        int LogicalStringCompare(const char16* string1, const char16* string2)
+        int LogicalStringCompare(const char16* string1, int str1size, const char16* string2, int str2size)
         {
-            return PlatformAgnostic::UnicodeText::Internal::LogicalStringCompareImpl(string1, string2);
```
We may want to have this definition be `#ifndef _WIN32` like GetLegacyCharacterClassificationType above, since even when we are using ICU on Windows, this function is trying to emulate CompareStringEx. This isn't an issue you introduced but it would be nice to fix it while this code is being touched. If you want to fix this, you will have to move the LogicalStringCompare definition in Platform\Windows\UnicodeText.cpp as well so it gets compiled in on Windows.

jackhorton commented on lib/Runtime/PlatformAgnostic/Platform/Windows/UnicodeText.cpp:
```diff
@@ -414,11 +414,11 @@ namespace PlatformAgnostic
+            // CompareStringEx called with these flags is equivalent to calling StrCmpLogicalW
             // and we have the added advantage of not having to link with shlwapi.lib just for one function
-            int i = CompareStringW(LOCALE_USER_DEFAULT, NORM_IGNORECASE | SORT_DIGITSASNUMBERS, string1, -1, string2, -1);
+            int i = CompareStringEx(LOCALE_NAME_INVARIANT, NORM_IGNORECASE | SORT_DIGITSASNUMBERS, string1, str1size, string2, str2size, NULL, NULL, 0);
```
Same LOCALE_NAME_USER_DEFAULT as noted above.

msdimitrov commented on lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.Common.cpp:
```diff
@@ -67,11 +68,20 @@ namespace PlatformAgnostic
+                    numLength++;
                 }
 
+                *len += numLength;
```
*len is accumulating because the string may contain other digits or characters before that, so `=` would just overwrite the value and would lead to unnecessary iterations overall.

msdimitrov commented:
Sorry for taking so long. I squashed the commits, and I think I've fixed the issue you mentioned with the emulation of CompareStringEx on Windows ICU.

jackhorton commented on bin/NativeTests/UnicodeTextTests.cpp:
```diff
@@ -84,4 +102,10 @@ namespace UnicodeTextTest
+
+    TEST_CASE("LogicalCompareString_EmbeddedNullCharacters", "[UnicodeText]")
+    {
+        TestNullCharacters(_u("\0\0ab\0\0c123def\0\0"), 15, _u("abc123def"), 9, 0);
```
I am honestly surprised that these should compare equal -- I just checked firefox and they agree, but I would have thought "\0" would be less than "a" and thus left should be less than right, _or_ left should be greater than right since its longer. Good test case!

jackhorton commented:
No worries on the delay, I have been busy doing infrastructure work so I probably wouldn't have seen any updates anyways. @dilijev or @MSLaguana want to take a look at this before I merge?

MSLaguana commented on lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.Common.cpp:
```diff
@@ -88,14 +98,18 @@ namespace PlatformAgnostic
                 Assert(p2 != nullptr);
 
-                while (*p1 && *p2)
+                int str1size = 0;
```
I find this name slightly confusing, since I was expecting that the size parameter would be fixed. It looks like it is only used to make sure we don't walk past the end of the strings, would it make sense to instead compute something like `const char16* str1End = p1 + p1size, *str2End = p2 + p2Size` and then just check `p1 < str1End && p2 < str2End` in the loop condition? That way we also don't accidentally bump either pointer without bumping the length-so-far.

msdimitrov replied:
Completely agree with you. Thanks!

dilijev commented:
Looking

dilijev commented on lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.Common.cpp:
```diff
@@ -170,6 +180,17 @@ namespace PlatformAgnostic
+                    {
+                        p2++;
+                    }
+
```
nit: extra blank line

jackhorton commented:
@digitalinfinity this fix impacts --no-icu and Lite builds of ChakraCore. Should it stay in master or go to 1.10?

dilijev commented:
I'd vote 1.10

dilijev commented:
Marked as `External Contributor` as a reminder that a Chakra team member needs to adopt, test, and merge this change.
